core/stats.go

```diff
@@ -57,24 +69,67 @@ func (ds *dockerService) ListContainerStats(
 		return nil, err
 	}
 
-	var mtx sync.Mutex
-	var wg sync.WaitGroup
-	var stats = make([]*runtimeapi.ContainerStats, 0, len(listResp.Containers))
-	for _, container := range listResp.Containers {
-		container := container
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			if containerStats, err := ds.getContainerStats(container.Id); err == nil && containerStats != nil {
-				mtx.Lock()
-				stats = append(stats, containerStats)
-				mtx.Unlock()
-			} else if err != nil {
-				logrus.Error(err, " Failed to get stats from container "+container.Id)
+	// Use pipeline pattern to collect stats for each container
+	// https://pkg.go.dev/golang.org/x/sync/errgroup#example-Group-Pipeline
+	g, ctx := errgroup.WithContext(ctx)
+	containerIDCh := make(chan string)
+
+	// send jobs
+	g.Go(func() error {
+		defer close(containerIDCh)
+		for _, container := range listResp.Containers {
+			select {
+			case containerIDCh <- container.Id:
+			case <-ctx.Done():
+				return ctx.Err()
 			}
-		}()
+		}
+		return nil
+	})
+
+	// limit concurrency to half of number of CPU
+	statsCh := make(chan *runtimeapi.ContainerStats)
+	numWorkers := runtime.NumCPU() / 2
+	if numWorkers == 0 {
+		numWorkers = 1
+	}
```

Comment:
```suggestion
  // we must have at least one worker
	numWorkers := math.Max(runtime.NumCPU() / 2, 1)
```

Reply:
Hi @neersighted , using `math.Max` needs to convert between integer and float64, it's boring. And I did not want to import a new package for just `max()`. In golang 1.21, there's a built-in function `max`. Shall we consider it?

Reply:
FWIW I also thought about suggesting this change but decided the original code kept things simple, didn't bring in other libs, and wont have any surprise behavior down the road :man_shrugging:

Reply:
Ah, that's fair. I'm fine with requiring go 1.21+, but cri-dockerd is compiled as part of MCR and therefore is subject to the limitations of the compiler toolchain.

cc @corhere WRT whether or not a `go 1.21` line in `go.mod` would be problematic.

Reply:
Requiring Go 1.21 would be problematic for MCR. I'm still working on migrating us to Go 1.20!